cdap-ui/app/cdap/components/PluginJSONCreator/Create/Content/WidgetCollection/WidgetAttributesCollection/WidgetAttributeInput/AttributeKeyValueInput/index.tsx

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import WidgetWrapper from 'components/ConfigurationGroup/WidgetWrapper';
+
+const AttributeKeyValueInput = ({
+  keyField,
+  valueField,
+  keyRequired,
+  valueRequired,
+  widgetToAttributes,
+  setWidgetToAttributes,
+  widgetID,
+  field,
+}) => {
+  const onKeyChange = (newVal) => {
+    setWidgetToAttributes((prevObjs) => ({
```

Comment:
not sure if you are doing this in the optimization, but these construction of the state object should happen at the context level, not at the individual component

Reply:
In the later PR, it updates widget attributes as following:
https://github.com/cdapio/cdap/blob/882eb810d3291ed6b39efaf6394a932584abb867/cdap-ui/app/cdap/components/PluginJSONCreator/Create/index.tsx#L121